source/common/upstream/outlier_detection_impl.cc

```diff
@@ -207,10 +233,58 @@ void DetectorImpl::onConsecutive5xxWorker(HostPtr host) {
   ejectHost(host, EjectionType::Consecutive5xx);
 }
 
+const double Utility::SR_STDEV_FACTOR = 1.9;
+
+double Utility::srEjectionThreshold(double sr_sum, std::vector<double>& sr_data) {
+  double mean = sr_sum / sr_data.size();
+  double stdev = 0;
+  std::for_each(sr_data.begin(), sr_data.end(),
+                [&stdev, mean](double& v) { stdev += std::pow(v - mean, 2); });
+  stdev /= sr_data.size();
+  stdev = std::sqrt(stdev);
+
+  return mean - (SR_STDEV_FACTOR * stdev);
+}
+
 void DetectorImpl::onIntervalTimer() {
   SystemTime now = time_source_.currentSystemTime();
+  std::unordered_map<HostPtr, double> valid_sr_hosts;
+  uint64_t significant_host_threshold = runtime_.snapshot().getInteger(
+      "outlier_detection.significant_host_threshold", config_.significantHostThreshold());
+  uint64_t rq_volume_threshold = runtime_.snapshot().getInteger(
+      "outlier_detection.rq_volume_threshold", config_.rqVolumeThreshold());
+  std::vector<double> sr_data;
+  double sr_sum = 0;
+  // reserve upper bound of vector size to avoid reallocation.
+  sr_data.reserve(host_sinks_.size());
```

Comment:
nit: can we do the check for hosts_sinks_.size() > significant_host_threshold here too. only reserve if it we need to.